homeassistant/components/modem_callerid/sensor.py

```diff
@@ -1,121 +1,139 @@
 """A sensor for incoming calls using a USB modem that supports caller ID."""
-import logging
+from __future__ import annotations
 
-from basicmodem.basicmodem import BasicModem as bm
 import voluptuous as vol
 
 from homeassistant.components.sensor import PLATFORM_SCHEMA, SensorEntity
+from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry
 from homeassistant.const import (
     CONF_DEVICE,
     CONF_NAME,
     EVENT_HOMEASSISTANT_STOP,
     STATE_IDLE,
 )
-import homeassistant.helpers.config_validation as cv
-
-_LOGGER = logging.getLogger(__name__)
-DEFAULT_NAME = "Modem CallerID"
-ICON = "mdi:phone-classic"
-DEFAULT_DEVICE = "/dev/ttyACM0"
-
-STATE_RING = "ring"
-STATE_CALLERID = "callerid"
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import config_validation as cv, entity_platform
+from homeassistant.helpers.typing import DiscoveryInfoType
+
+from .const import (
+    DATA_KEY_API,
+    DEFAULT_DEVICE,
+    DEFAULT_NAME,
+    DOMAIN,
+    ICON,
+    SERVICE_REJECT_CALL,
+    STATE_CALLERID,
+    STATE_RING,
+)
 
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {
-        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
-        vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string,
-    }
+PLATFORM_SCHEMA = cv.deprecated(
+    vol.All(
+        PLATFORM_SCHEMA.extend(
+            {
+                vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+                vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string,
+            }
+        )
+    )
 )
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up modem caller ID sensor platform."""
+async def async_setup_platform(
+    hass: HomeAssistant,
+    config: ConfigEntry,
+    async_add_entities: entity_platform.AddEntitiesCallback,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> None:
+    """Set up the Modem Caller ID component."""
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=config
+        )
+    )
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: entity_platform.AddEntitiesCallback,
+) -> None:
+    """Set up the Modem Caller ID sensor."""
+    api = hass.data[DOMAIN][entry.entry_id][DATA_KEY_API]
+    async_add_entities(
+        [
+            ModemCalleridSensor(
+                api,
+                entry.data.get(CONF_NAME),
+                entry.data[CONF_DEVICE],
+                entry.entry_id,
+            )
+        ],
+        True,
+    )
+
+    @callback
+    async def _async_on_hass_stop(self, _):
```

Comment:
```suggestion
    def _async_on_hass_stop(self, _):
```

It can't be a callback and a coroutine. Since nothing is being awaited, lets make it just a callback